DSL rules (interacting with items) created via the UI result in errors

I am running OH3 build #1945.
When I write a simple rule and save this as a .rules file in conf/rules this rule is properly executed.
When I create the same rule in the UI, the rule does not execute. I get the following in the log:

[ERROR] [internal.handler.ScriptActionHandler] - Script execution failed: An error occurred during the script execution: null

My file based rule, when looking in the UI in the 'code' view of the rule is shown as:

```
triggers:

id: “0”
configuration:
cronExpression: 0 0 0 * * ?
type: timer.GenericCronTrigger
conditions: []
actions:
inputs: {}
id: script
configuration:
type: application/vnd.openhab.dsl.rule
script: |
// context: system-2
counter = Uptime.state as DecimalType
counter = counter + 1
postUpdate(Uptime, counter)
watchdog = 0
postUpdate(Watchdog, watchdog)
type: script.ScriptAction
```

The rule created in UI:

```
triggers:

id: “0”
configuration:
cronExpression: 0 0 0 * * ?
type: timer.GenericCronTrigger
conditions: []
actions:
inputs: {}
id: script
configuration:
type: application/vnd.openhab.dsl.rule
script: |
var Number counter = Uptime.state as DecimalType
counter = counter + 1
postUpdate(Uptime, counter)
var Number watchdog = 0
postUpdate(Watchdog, watchdog)
type: script.ScriptAction
```

The main difference is that in my file based rules, I have declared the variables globally. In the UI based rule I declared these in the rule.

When I copy the line with // context to the UI based rule, the rule is executed and no errors are shown in the log.

I have done some more tests on a clean install of OH3 build #1945, running in a Docker container.
The setup is as follows:
- created 1 item named Uptime via the UI
- created 1 rule named  system.rules with content:

```
var Number counter

rule "Count days up"
when
    Time cron "0 0 0 * * ?"
then
    logInfo("runtime.busevents", "Increase uptime")
    counter = Uptime.state as DecimalType
    counter = counter + 1
    postUpdate(Uptime, counter)
end
```

- created 1 rule via the UI with content:

```
logInfo("runtime.busevents", "Increase uptime")
var Number counterTest = Uptime.state as DecimalType
counterTest = counterTest + 1
postUpdate(Uptime, counterTest)
```

The view in 'code' tab page of this UI created rule is this:

```
triggers:
  - id: "1"
    configuration:
      cronExpression: 0 0 0 * * ? *
    type: timer.GenericCronTrigger
conditions: []
actions:
  - inputs: {}
    id: "2"
    configuration:
      type: application/vnd.openhab.dsl.rule
      script: |-
        logInfo("runtime.busevents", "Increase uptime")
        var Number counterTest = Uptime.state as DecimalType
        counterTest = counterTest + 1
        postUpdate(Uptime, counterTest)
    type: script.ScriptAction
```

The file based rule executes normally, the Uptime item is incremented when I trigger the rule manually (by clicking on the 'run now' play button). Also in the openhab.log I see the logInfo:

`2020-10-04 10:41:39.304 [INFO ] [.core.model.script.runtime.busevents] - Increase uptime`

Then, when I execute the UI based rule, the Uptime item is not incremented, and in the openhab.log I notice:
```
2020-10-04 10:50:43.275 [INFO ] [.core.model.script.runtime.busevents] - Increase uptime
2020-10-04 10:50:43.276 [ERROR] [internal.handler.ScriptActionHandler] - Script execution failed: An error occurred during the script execution: null
```

So, the logInfo is actually executed.
When I selectively comment out some of the lines in my rule, I notice that the null error only comes as soon as I have a line which is reading or writing an Item.

When I reduce the rule, this gives no errors:
```
var Number counterTest = 2
logInfo("runtime.busevents", "Increase uptime")
logInfo("runtime.busevents", "Uptime: "+ counterTest)
```

And a rule with only this line gives the same null error:
`postUpdate(Uptime, 2)`

Also this single line rule gives the same null error:
`var Number counterTest = Uptime.state as DecimalType`


Hope this helps, if there is anything more I can test, let me know.

best regards
Bram

--- a/bundles/org.openhab.core.model.script.runtime/src/org/openhab/core/model/script/runtime/internal/engine/DSLScriptEngine.java
+++ b/bundles/org.openhab.core.model.script.runtime/src/org/openhab/core/model/script/runtime/internal/engine/DSLScriptEngine.java
@@ -23,11 +23,15 @@ import javax.script.ScriptEngineFactory;
 import javax.script.ScriptException;
 import javax.script.SimpleScriptContext;
 
+import org.eclipse.emf.ecore.resource.Resource;
 import org.eclipse.jdt.annotation.Nullable;
 import org.eclipse.xtext.naming.QualifiedName;
+import org.eclipse.xtext.resource.IResourceServiceProvider;
+import org.eclipse.xtext.resource.XtextResource;
 import org.eclipse.xtext.xbase.XExpression;
 import org.eclipse.xtext.xbase.interpreter.IEvaluationContext;
 import org.eclipse.xtext.xbase.interpreter.impl.DefaultEvaluationContext;
+import org.openhab.core.model.script.engine.Script;
 import org.openhab.core.model.script.engine.ScriptExecutionException;
 import org.openhab.core.model.script.engine.ScriptParsingException;
 import org.openhab.core.model.script.runtime.DSLScriptContextProvider;
@@ -110,7 +114,7 @@ public class DSLScriptEngine implements javax.script.ScriptEngine {
             } else {
                 s = scriptEngine.newScriptFromString(script);
             }
-            IEvaluationContext evalContext = createEvaluationContext(specificContext);
+            IEvaluationContext evalContext = createEvaluationContext(s, specificContext);
             s.execute(evalContext);
         } catch (ScriptExecutionException | ScriptParsingException e) {
             throw new ScriptException(e.getMessage(), modelName, -1);
@@ -118,8 +122,20 @@ public class DSLScriptEngine implements javax.script.ScriptEngine {
         return null;
     }
 
-    private DefaultEvaluationContext createEvaluationContext(IEvaluationContext specificContext) {
-        DefaultEvaluationContext evalContext = new DefaultEvaluationContext(specificContext);
+    private DefaultEvaluationContext createEvaluationContext(Script script, IEvaluationContext specificContext) {
+        IEvaluationContext parentContext = specificContext;
+        if (specificContext == null && script instanceof ScriptImpl) {
+            XExpression xExpression = ((ScriptImpl) script).getXExpression();
+            if (xExpression != null) {
+                Resource resource = xExpression.eResource();
+                IEvaluationContext evaluationContext = null;
+                if (resource instanceof XtextResource) {
+                    IResourceServiceProvider provider = ((XtextResource) resource).getResourceServiceProvider();
+                    parentContext = provider.get(IEvaluationContext.class);
+                }
+            }
+        }
+        DefaultEvaluationContext evalContext = new DefaultEvaluationContext(parentContext);
         for (Map.Entry<String, String> entry : implicitVars.entrySet()) {
             Object value = context.getAttribute(entry.getKey());
             if (value != null) {

